### PR TITLE
Ability to start OS Dashboards with a newer and compatible nodejs version 

### PR DESCRIPTION
**Pull request by sipopo**
### Description
Resolve bug the application doesn't start with nodejs version 14.20.0
 
### Issues Resolved
[BUG] Doesn't start with nodejs 14.20.0 #2088

### CODE REVIEW

**Review by kavilla — DISMISSED** [review]
Added details to https://github.com/opensearch-project/OpenSearch-Dashboards/issues/2088, this logic is intentional.

File: src/setup_node_env/node_version_validator.js
Comment (smortex) [inline src/setup_node_env/node_version_validator.js:47]: The error message bellow needs to be updated accordingly: a higher minor version is fine too.

Comment (sipopo) [reply src/setup_node_env/node_version_validator.js:48]: I added warning message.

**Comment by sipopo** [comment]
I added warning message for minor check

File: src/setup_node_env/node_version_validator.js
Comment (kavilla) [reply src/setup_node_env/node_version_validator.js:48]: > The error message bellow needs to be updated accordingly: a higher minor version is fine too.

Got it makes sense

**Comment by sipopo** [comment]
Hi, It is my first pull request in github. Do I need something to do else?

Sorry for my question...

**Comment by smortex** [comment]
#2101 was merged and update nodejs to 14.20.0, but the point of this PR/issue #2088 remains.  I would suggest clarifying this by updating this Issue and PR subject to something like "Allow using a newer minor version of nodejs than the bundled one" or "Fix startup when using a newer minor version of nodejs then the bundled one" (since the issue is not 14.20.0 specific anymore).

At the bottom of the page, you will see 2 failing tests:

screenshot

Click "Details" to see what was wrong, fix it and update the PR accordingly (see bellow for some hints).  Each project is different, but the idea is usually to external contributors to work on their code and have automatic feedback, and reviewers will wait for all tests to be green before considering merging code.  Until then, they may still of course help you.

You mention this is your first PR (awesome!), so it's good to know that any update you do in your branch will update this PR: you can re-write your commits and force push, and the PR will automagically update!


For example assuming you want to move your commits on top of master, sign them all to make the "Developer Certificate of Origin" check green, and rework them in a single commit (squash them), you can, from your working branch:

```
git fetch origin       # Download the latest code we have here
git rebase origin/main --signoff --interactive # Interactively move your commits on top of the main branch ensuring they are all signed
```

This will bring your editor listing all commits in your branch.  Keep the first line unchanged, and replace `pick` with `squash` for all subsequent lines.  Save and quit, your editor will pop again and allow you to combine all the commit messages in a single one.

Then push your updated branch:

```
git push -f            # Send the changes (-f is required because we re-wrote history)
```

**Comment by ashwin-pc** [comment]
@sipopo will you be able to add a signoff to your change? We cannot accept your change unless we have that. The PR otherwise looks great! @smortex's comment has detailed instructions on how to do so

**Comment by sipopo** [comment]
Sorry for late. I signed my commit.

**Comment by sipopo** [comment]
Sorry for my misunderstanding. I returned my changes to project )

**Comment by ashwin-pc** [comment]
@sipopo Thanks for making the change!

**Comment by ashwin-pc** [comment]
@sipopo Looks like your change is breaking a version check test.

**Comment by joshuarrrr** [comment]
We'll update this issue to make sure the community is clear on the change: https://github.com/opensearch-project/OpenSearch-Dashboards/issues/2088

But otherwise, @sipopo we agree with these changes and will plan to merge once the tests are updated.

@KrooshalUX Can you approve the wording of the warning.

**Comment by kgcreative** [comment]
> We'll update this issue to make sure the community is clear on the change: #2088
> 
> But otherwise, @sipopo we agree with these changes and will plan to merge once the tests are updated.
> 
> @KrooshalUX Can you approve the wording of the warning.

Pulling the warning message out of the files for easier parsing:

```
var warnMessage =
    `OpenSearch Dashboards was built with ${requiredVersion}. ` +
    `The current Minor version of Node.js ${currentVersion} didn't test completely.`;
  console.warn(warnMessage);
```
Parsed with an example below:
`OpenSearch Dashboards was built with 14.20.0. The current Minor version of Node.js 14.20.1 didn't test completely.`

The language of this feels a little unclear in terms of next steps and remediation. "Didn't test completely" doesn't tell me what went wrong or how to fix.

Consider:
`OpenSearch Dashboards requires NodeJS 14.20.0. The current version is 14.20.1`

Can we get a little more clarity around what "Didn't test completely" means? Are there partial failures? Partial successes?

Perhaps `Some tests failed due to a Node.JS version mismatch. OpenSearch Dashboards requires NodeJS ${currentVersion} . The current version is ${currentVersion}.

**Comment by joshuarrrr** [comment]
Yeah, "didn't test completely" just means "this _might_ work, but no guarantees".

Maybe

`You're using an untested version of Node.js. OpenSearch Dashboards has been tested to work with Node.js ${requiredVersion}. Your current version is ${currentVersion}.`

**Comment by kgcreative** [comment]
> `You're using an untested version of Node.js. OpenSearch Dashboards has been tested to work with Node.js ${requiredVersion}. Your current version is ${currentVersion}.`

This works for me

**Comment by joshuarrrr** [comment]
Thanks, @sipopo, I'll plan to take a look at the latest version on Monday!

**Comment by ashwin-pc** [comment]
@sipopo Thanks for the change, it looks good. The changelog check is failing because we introduced a changelog requirement for all PR's. Its a simple requirement. You just need to add an entry in the `./Changelog.md` file for your change. Since this has been open for a while and since everything else looks good I'm okay merging it in as is and adding the changelog afterwards. But if possible, it would be ideal if the changelog entry happens in the same PR as the change itself for future reference.

**Comment by sipopo** [comment]
@ashwin-pc  Could you give me text for Changelog.md, and I will add it. My english is not so perfect :)

**Comment by ashwin-pc** [comment]
You can add the title of the original issue
```
* Ability to start OS Dashboards with a newer and compatible nodejs version ([#2091](https://github.com/opensearch-project/OpenSearch-Dashboards/pull/2091))
```

You can add this under the Unreleased -> Features/Enhancements section

**Review by ashwin-pc — DISMISSED** [review]
Thanks for sticking with the change @sipopo. LGTM!

**Comment by ashwin-pc** [comment]
@kavilla @joshuarrrr Can you take a look at this? We can fix the conflict ourselves in github without dismissing the approvals.

File: src/setup_node_env/node_version_validator.js
Comment (AMoo-Miki) [inline src/setup_node_env/node_version_validator.js:54]: If `requiredVersion` was `14.20.20` and `currentVersion` was `14.20.3`, `isVersionValid` would be true and `isMinorValid` would also be true; this should not be accepted. We cannot use `parseFloat`.

A better option would be:

```suggestion
var version = {
    current: {
        major: currentVersionMajorMinorPatch[1],
        minor: parseInt(currentVersionMajorMinorPatch[2], 10),
        patch: parseInt(currentVersionMajorMinorPatch[3], 10),
    },
    required: {
        major: requiredVersionMajorMinorPatch[1],
        minor: parseInt(requiredVersionMajorMinorPatch[2], 10),
        patch: parseInt(requiredVersionMajorMinorPatch[3], 10),
    },
}

var isVersionValid =
  version.current.major === version.required.major &&
  version.current.minor === version.required.minor &&
  version.current.patch >= version.required.patch;

var isMinorValid =
    version.current.major === version.required.major &&
    version.current.minor > version.required.minor;
```

and then the condition below would be
```js
!isVersionValid && !isMinorValid
```

**Review by AMoo-Miki — CHANGES_REQUESTED** [review]

File: src/setup_node_env/node_version_validator.js
Comment (AMoo-Miki) [inline src/setup_node_env/node_version_validator.js:61]: With my proposal, this should be
```suggestion
} else if (isMinorValid) {
```

... because `isMinorValid` really means versions were not good at the patch level but are good at the minor level.

**Comment by joshuarrrr** [comment]
@sipopo Will you be able to review @AMoo-Miki's suggestions and push updated changes?

File: src/setup_node_env/node_version_validator.js
Comment (sipopo) [reply src/setup_node_env/node_version_validator.js:54]: @AMoo-Miki  thank you for advice

Comment (sipopo) [reply src/setup_node_env/node_version_validator.js:54]: It's strange but I couldn't accept your suggestions.

Comment (sipopo) [reply src/setup_node_env/node_version_validator.js:54]: I see this error "Applying suggestions on deleted lines is currently not supported. "

**Comment by kavilla** [comment]
> @sipopo Will you be able to review @AMoo-Miki's suggestions and push updated changes?

I think @sipopo has been playing ball and I think this we should move forward. Do we want to consider making the suggested changes and using @sipopo's author commits as a base to get them as a contributor?

**Comment by sipopo** [comment]
Hi, Are you waiting for me?

**Comment by ashwin-pc** [comment]
@sipopo yes, can you incorporate @AMoo-Miki 's changes in your PR manually? @kavilla was looking to see if someone could pick it up instead, but it looks like no one has the bandwidth right now.

**Comment by ashwin-pc** [comment]
So i pulled down the changes to validate if this works as expected only to run into a very interesting behavior. @sipopo @AMoo-Miki How did the two of you validate this functionality? Because when i run the yarn commands, it turns out that the first line of defense is node and package.json itself. 

We have an `engines` key in the [package.json](https://github.com/opensearch-project/OpenSearch-Dashboards/blob/main/package.json#L465) file that fails first when we try to use any other minor or patch version of node. It does not even get to node check. The only two ways to get past this is to update the engine requirements to be something like `^14.20.0` or pass the `--ignore-engines` flag. Did either of you run into this while testing the feature? Also if you use the former option of specifying the node version using the caret symbol, the regex that you use is no longer valid. Here's one way to fix that:
1. Delete line 37 where we prefix `v` and rename `rawRequiredVersion` to `requiredVersion`. We know that this will not be null for our usecases so this is a safe assumption. And to be honest, i'm not sure that line makes any sense.
2. Modify the version regex to `/^[v~=\^]+(\d+)\.(\d+)\.(\d+)/` to handle `v`, `^`, `~` and `=` prefixes 

With these changes, the rest of the code works as expected. I think this is also why @kavilla and @miki might have called out the initial change did not work.

File: CHANGELOG.md
Comment (ashwin-pc) [inline CHANGELOG.md:1]: Can you remove this change?

**Comment by AMoo-Miki** [comment]
> So i pulled down the changes to validate if this works as expected only to run into a very interesting behavior. @sipopo @AMoo-Miki How did the two of you validate this functionality? Because when i run the yarn commands, it turns out that the first line of defense is node and package.json itself.

The original idea was that this piece would relax the requirement and then `engines.node` would use `14.20`. I felt that didn't eliminate the restrictions enough and hence I implemented this very differently in https://github.com/opensearch-project/OpenSearch-Dashboards/pull/3402.

@sipopo I would have loved to add you as a co-author but due to the missing DCO in this PR, I couldn't. Thanks for all the effort you put in here; know that you were a big part of the motivation to get this done.

**Pull request closed without merging**
